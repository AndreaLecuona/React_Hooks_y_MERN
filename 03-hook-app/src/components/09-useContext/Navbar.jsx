import React from 'react';
import { Link, NavLink } from 'react-router-dom';

export const Navbar = () => {
    return (
        <nav>
            <ul>
                <h2>
                    Learning React Router + useContext
                </h2>
                <li>
                    <NavLink exact to="./" activeClassName="isActive" >Home</NavLink>
                </li>
                <li>
                    <NavLink exact to="./about" activeClassName="isActive">About</NavLink>
                </li>
                <li>
                    <NavLink exact to="./login" activeClassName="isActive" >Login</NavLink>
                </li>
            </ul>
        </nav>
    )
}

// si usaramos anchor tags con hrefs, la pagina completa se renderiza cada vez que cambio de pagina.
// el <Link> es el equivalente del <a>, y el 'to' es el equivalente del 'href'. No necesita renderizar
//      la pagina completa cunado cambio de pagina y eso ahorra tiempo de proceso

// la diferencia entre <Link> y <Navlink> es que Navlink permite tener el atributo "activeClassName" para
//      darle estilo a la pagina que coincide con el path, es decir, la que ESTÁ ACTIVA


// export const Navbar = () => {
//     return (
//         <nav>
//             <ul>
//                 <li>
//                     <Link to="./">Home</Link>
//                 </li>
//                 <li>
//                     <Link to="./about">About</Link>
//                 </li>
//                 <li>
//                     <Link to="./login">Login</Link>
//                 </li>
//             </ul>
//         </nav>
//     )
// }