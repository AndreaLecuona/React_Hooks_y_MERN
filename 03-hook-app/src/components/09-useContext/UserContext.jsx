import { createContext } from 'react';

export const UserContext = createContext(null);

// Es un higher order component. 
// Es un espacio independiente que contiene cierta informaci√≥n, al que pueden recurrir todos los componentes que configuren su acceso.
// Sirve para transmitir informacion (o props) entre elementos que NO POSEEN RELACION PADRE-HIJO.
