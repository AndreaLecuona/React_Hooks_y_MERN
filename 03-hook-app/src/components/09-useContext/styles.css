@import url('https://fonts.googleapis.com/css2?family=Mitr:wght@500&family=Pridi&display=swap');

:root{
    --nav-font: 'Mitr', sans-serif;
    --nav-weight: 500;
    --title-font: 'Pridi', serif;
    --title-weight: 400;

    --primary-pink:#f5c3c1;
    --primary-blue:#4a6279;

    --accent: #f8f687;

    --neutral-dark: #373847;
}

.wrapper{
    width: 90%;
    margin: 0 auto;
    color: var(--neutral-dark);
}
nav{
    background-color: var(--primary-blue);
    border-radius: 0 0 10px 10px;
}
nav ul{
    display: flex;
    justify-content: space-around;
    padding: 0;
}
nav ul li{
    list-style: none;
    padding: 1rem 0;
    transition: all 0.3s ease-in-out;
}
nav ul li:hover{
    background-color: var(--primary-pink);
}
nav ul li a{
    text-decoration: none;
    color: var(--primary-pink);
    font-family: var(--nav-font);
    font-weight: var(--nav-weight);
    font-size: 1rem;
    transition: all 0.3s ease-in-out;
    padding: 1rem;
}
nav ul li a:hover{
    color: var(--primary-blue);
}

h2{
    font-size: 1rem;
    color: white;
    padding: 1rem 0;
    margin: auto 0;
    width: 25%;
    font-family: var(--title-font);
    font-weight: var(--title-weight);
}

.isActive{
    color: var(--accent);
}


h1{
    font-family: var(--title-font);
    font-weight: var(--title-weight);
    margin-top: 4rem;
}

.btn{
    border: 3px dashed var(--primary-blue);
    background-color: var(--primary-pink);
    color: var(--primary-blue);
    padding: 0.5rem 2rem;
    font-family: var(--nav-font);
    border-radius: 50px;
}
.btn:hover{
    border: 3px dashed var(--primary-pink);
    background-color: var(--primary-blue);
    color: var(--primary-pink);
}