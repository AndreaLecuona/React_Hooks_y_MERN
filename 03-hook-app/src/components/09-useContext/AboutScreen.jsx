import React, { useContext } from 'react'
import { UserContext } from './UserContext';

export const AboutScreen = () => {

    const { user, setUser } = useContext( UserContext );

    const handleClick = () => {
        setUser( {} )
    }

    return (
        <div>
            <h1>About screen</h1>
            <hr/>

            <pre>
                { JSON.stringify( user, null, 3 ) }
            </pre>
            
            <button 
            className="btn"
            onClick={ handleClick }
            >
                Purge user's data in all pages
            </button>
        </div>
    )
}
