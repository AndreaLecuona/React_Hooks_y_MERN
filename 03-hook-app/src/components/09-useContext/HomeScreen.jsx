import React, { useContext } from 'react'
import { UserContext } from './UserContext'

export const HomeScreen = () => {

    const { user, setUser } = useContext(UserContext);

    // console.log(user.city);

    return (
        <div>
            <h1>Home screen</h1>
            <hr/>

            <pre>
                { JSON.stringify( user, null, 3 ) }
            </pre>
        </div>
    )
}

// Para que RECIBA LA INFORMACION, al useContext le pasamos como argumento el higher-order component.
// podemos desesctructurar el valor resultado para ordenar y utilizar mejor la info que necesito

// el home solo usa el 'user' porque quiere MOSTRAR la info
// el login solo usa el 'setUser' porque quiere MODIFICAR la info