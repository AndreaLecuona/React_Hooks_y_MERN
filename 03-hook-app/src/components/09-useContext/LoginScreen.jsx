import React, { useContext } from 'react'
import { UserContext } from './UserContext';

export const LoginScreen = () => {

    const newUser = {
        name:'Andre',
        city:'Buenos Aires',
        email:'email@email.com'
    }

    const { user, setUser } = useContext( UserContext );

    return (
        <div>
            <h1>Login screen</h1>
            <hr/>
            <button 
            className="btn"
            onClick={ () => setUser(newUser) }
            >
                Show user's info in all pages
            </button>
        </div>
    )
}

// Para que RECIBA LA INFORMACION, al useContext le pasamos como argumento el higher-order component.
// podemos desesctructurar el valor resultado para ordenar y utilizar mejor la info que necesito

// el home solo usa el 'user' porque quiere MOSTRAR la info
// el login solo usa el 'setUser' porque quiere MODIFICAR la info
