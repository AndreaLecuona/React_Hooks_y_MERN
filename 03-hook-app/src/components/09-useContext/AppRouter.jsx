import React from 'react'
import {
    BrowserRouter as Router,
    Switch,
    Route,
    Redirect
} from "react-router-dom";
import { Navbar } from './Navbar';
import { AboutScreen } from './AboutScreen';
import { HomeScreen } from './HomeScreen';
import { LoginScreen } from './LoginScreen';

export const AppRouter = () => {
    return (
        <Router>
            <div>

            <Navbar/>

            <Switch>
                <Route exact path="/" component={ HomeScreen }/>
                <Route exact path="/about" component={ AboutScreen }/>
                <Route exact path="/login" component={ LoginScreen }/>

                {/* <Route component={ HomeScreen }/> */}
                <Redirect to="/"/>
            </Switch>

            </div>
        </Router>
    )
}

// El switch renderiza rutas de manera condicional, que en este caso tienen jerarquía de hermanos

//con sólo este router, puedo navegar cambiando la terminacion del path en el url
// si el path no existe, entonces puedo poner un Route SIN PATH como default e indicarle qué quiero renderizar (en este caso es la HomeScreen pero podria ser una 404)
// otra forma para enfrentar path que no existen, es usar el Redirect

// SI NO UTILIZO LA CONDICIÓN EXACT, entonces el home screen (o ruta más general) siempre se 
// pone al final de la lista porque el Route devuelve la PRIMER coincidencia 
// con el path. Si lo pusiera al comienzo, no podría llegar a los otros paths 
// porque siempre la primer coincidencia sería el home