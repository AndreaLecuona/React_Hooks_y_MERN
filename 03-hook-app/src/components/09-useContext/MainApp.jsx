import React, { useState } from 'react'
import { AppRouter } from './AppRouter'
import { UserContext } from './UserContext'
import './styles.css'

export const MainApp = () => {

    const [user, setUser] = useState({});

    return (
        <div className="wrapper">
            <UserContext.Provider value={ {
                user: user, 
                setUser: setUser
                } }>

                <AppRouter/>

            </UserContext.Provider>
        </div>
    )
}

// Agregar el 'Provider' habilita a compartir la informacion entre hijos
// El 'value' es la informacion que voy a compartir a través del useContext que funciona dentro del higher-order component llamado 'UserContext'
// en este caso comienza como un objeto vacío porque lo va a 'llenar' el 'login' y el resultado se va a mostrar en el 'home'