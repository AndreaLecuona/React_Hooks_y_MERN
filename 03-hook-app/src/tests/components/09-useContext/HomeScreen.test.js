import React from 'react';
import { mount } from 'enzyme';
import { HomeScreen } from '../../../components/09-useContext/HomeScreen';
import { UserContext } from '../../../components/09-useContext/UserContext';

describe('Pruebas en HomeScreen', () => {

    const user = { 
        name: 'Ziggy',
        email: 'stardust@spiders.mars'
    }
    
    //si usaramos shallow, solo renderizaría el UserContext con el HomeScreen vacío, NO renderizaría los elementos internos del homeScreen (div, h1, pre, etc.)
    const wrapper = mount( 
        <UserContext.Provider value={ { user } }>
        <HomeScreen/>
        </UserContext.Provider>
    );

    test('debe renderizarse correctamente', () => {

        //si no hubieramos puesto el provider en el wrapper general, esta prueba fallaría porque no tendria acceso a los valores del context que este componente usa
        expect( wrapper ).toMatchSnapshot();

    })
    
})
