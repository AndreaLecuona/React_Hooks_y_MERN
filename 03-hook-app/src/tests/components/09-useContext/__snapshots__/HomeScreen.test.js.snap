// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en HomeScreen debe renderizarse correctamente 1`] = `
<div>
  <h1>
    Home screen
  </h1>
  <hr />
  <pre>
    {
   "name": "Ziggy",
   "email": "stardust@spiders.mars"
}
  </pre>
</div>
`;
